### DISCUSSION TITLE
[Query] What is the effect of fsfreeze pre-hook on running application ?

### DISCUSSION THREAD

**Discussion by navilg** [vmware-tanzu/velero#6313] [Community support Q&A] [answered]
Hello All,

Hope you are doing well.

I am using velero 1.8.1 with restic integration. And I have PV on NFS server. 
I wanted to understand, how fsfreeze pre-hook affect the running application ? 

When a filesystem is freezed, WIll it allow application to work properly (like read and write operation on filesystem by app pods during backup) ?
What is the best way of using it ? I am trying not to have service degradation during backups.

Thanks

**Comment by navilg** [comment]
@Lyndon-Li Can I please get your attention on this query

**Comment by Lyndon-Li** [comment, marked as answer]
@navilg 
> When a filesystem is freezed, WIll it allow application to work properly

Most probably not, the kernel will hold the write IO once `fsfreeze --freeze` is called, until `fsfreeze --unfreeze`. It means the application cannot make any write call to the file system, otherwise, it will hang (in most cases).

> What is the best way of using it

It depends on backup types.
Let's assume you call `fsfreeze --freeze` in pre-hook and call `fsfreeze --unfreeze` in post-hook.
If the target is not making big impact to the application during backup, then:

- If you are using file system backup, the advice is, never use the hook. Otherwise, the file system will be frozen during the entire file system backup. In another words, if the file system backup takes 1 hour, the file system will be frozen for 1 hour. Obviously, the impact to the application is very large
- If you are using CSI snapshot backup, it is OK to use the hook. The file system is frozen when snapshot creation starts and unfrozen after snapshot is created. The time lag should be very short.

For sure, if you want to move the data to a 2nd place and by the mean time, you want to use the hook to get a better consistency data state, there is not this way at present. 
From v1.12, Velero will allow users to move the CSI snapshot data to a 2nd place, then users will be able to take CSI snapshot with the hook and move the data to a 2nd place.

**Reply by navilg** [reply]
Thanks for clarifying @Lyndon-Li

**Comment by caijinhu123** [comment]
"Hello, I have a question. When I was testing the hook, I used the example provided by the official website. After executing the sfreeze --freeze command in the fsfreeze container, I got a permission denied error. This caused errors when taking CSI snapshot backups. Can you explain why?"@Lyndon-Li
